src/mapgen_functions.cpp

```diff
@@ -3779,18 +3779,34 @@ void mapgen_forest( map *m, oter_id terrain_type, mapgendata dat, const time_poi
         m->draw_fill_background( current_biome_def.groundcover );
     }
 
+    // There is a chance of placing terrain dependent furniture, e.g. f_cattails on t_water_sh.
+    const auto get_terrain_dependent_furniture = [&current_biome_def]( const ter_id tid ) -> const
+    furn_id * {
+        const auto terrain_dependent_furniture_it = current_biome_def.terrain_dependent_furniture.find( tid );
+        if( terrain_dependent_furniture_it == current_biome_def.terrain_dependent_furniture.end() )
+        {
+            return nullptr;
+        }
+
+        const forest_biome_terrain_dependent_furniture tdf = terrain_dependent_furniture_it->second;
+        if( one_in( tdf.chance ) )
+        {
+            const auto fid = tdf.furniture.pick();
+            return fid;
+        }
+
+        return nullptr;
+    };
```

Comment:
Given how this is used we can simplify the process a bit by moving the furniture setting into the lambda itself.
Should be able to modify the lambda to the following:
```
const auto set_terrain_dependent_furniture = [&current_biome_def, &m]( const ter_id tid, const int x, const int y ) -> void {
    const auto terrain_dependent_furniture_it = current_biome_def.terrain_dependent_furniture.find( tid );
    // If tid doesn't have any terrain dependent furnitures we cannot continue
    if( terrain_dependent_furniture_it != current_biome_def.terrain_dependent_furniture.end() ) {
        const forest_biome_terrain_dependent_furniture tdf = terrain_dependent_furniture_it->second;
        // If weight is zero we will get a nullptr result if this goes through, so catch that before doing 1 in N test
        if( ( tdf.furniture.get_weight() > 0 ) && one_in( tdf.chance ) )
        {
            const auto fid = tdf.furniture.pick();
            m->furn_set( x, y, *fid );
        }
    }
};
```
and call in the loop using `set_terrain_dependent_furniture( feature.ter, x, y );`

Reply:
I'm on board with it, though I went with a slightly different implementation--I prefer the early return to nested conditionals.